File: besu/src/main/java/org/hyperledger/besu/controller/MergeBesuControllerBuilder.java
/*
 * Copyright contributors to Hyperledger Besu.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package org.hyperledger.besu.controller;

import org.hyperledger.besu.consensus.merge.MergeContext;
import org.hyperledger.besu.consensus.merge.MergeProtocolSchedule;
import org.hyperledger.besu.consensus.merge.PostMergeContext;
import org.hyperledger.besu.consensus.merge.TransitionBestPeerComparator;
import org.hyperledger.besu.consensus.merge.blockcreation.MergeCoordinator;
import org.hyperledger.besu.datatypes.Address;
import org.hyperledger.besu.datatypes.Hash;
import org.hyperledger.besu.ethereum.ProtocolContext;
import org.hyperledger.besu.ethereum.blockcreation.MiningCoordinator;
import org.hyperledger.besu.ethereum.chain.Blockchain;
import org.hyperledger.besu.ethereum.core.BlockHeader;
import org.hyperledger.besu.ethereum.core.Difficulty;
import org.hyperledger.besu.ethereum.core.MiningConfiguration;
import org.hyperledger.besu.ethereum.eth.EthProtocolConfiguration;
import org.hyperledger.besu.ethereum.eth.manager.EthContext;
import org.hyperledger.besu.ethereum.eth.manager.EthMessages;
import org.hyperledger.besu.ethereum.eth.manager.EthPeers;
import org.hyperledger.besu.ethereum.eth.manager.EthProtocolManager;
import org.hyperledger.besu.ethereum.eth.manager.EthScheduler;
import org.hyperledger.besu.ethereum.eth.manager.MergePeerFilter;
import org.hyperledger.besu.ethereum.eth.manager.peertask.PeerTaskExecutor;
import org.hyperledger.besu.ethereum.eth.peervalidation.PeerValidator;
import org.hyperledger.besu.ethereum.eth.peervalidation.RequiredBlocksPeerValidator;
import org.hyperledger.besu.ethereum.eth.sync.SynchronizerConfiguration;
import org.hyperledger.besu.ethereum.eth.sync.backwardsync.BackwardChain;
import org.hyperledger.besu.ethereum.eth.sync.backwardsync.BackwardSyncContext;
import org.hyperledger.besu.ethereum.eth.sync.state.SyncState;
import org.hyperledger.besu.ethereum.eth.transactions.TransactionPool;
import org.hyperledger.besu.ethereum.forkid.ForkIdManager;
import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
import org.hyperledger.besu.ethereum.mainnet.ScheduleBasedBlockHeaderFunctions;
import org.hyperledger.besu.ethereum.worldstate.WorldStateArchive;

import java.util.List;
import java.util.Optional;
import java.util.OptionalLong;
import java.util.concurrent.atomic.AtomicReference;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/** The Merge besu controller builder. */
public class MergeBesuControllerBuilder extends BesuControllerBuilder {
  private final AtomicReference<SyncState> syncState = new AtomicReference<>();
  private static final Logger LOG = LoggerFactory.getLogger(MergeBesuControllerBuilder.class);

  /** Default constructor. */
  public MergeBesuControllerBuilder() {}

  @Override
  protected MiningCoordinator createMiningCoordinator(
      final ProtocolSchedule protocolSchedule,
      final ProtocolContext protocolContext,
      final TransactionPool transactionPool,
      final MiningConfiguration miningConfiguration,
      final SyncState syncState,
      final EthProtocolManager ethProtocolManager) {
    return createTransitionMiningCoordinator(
        protocolSchedule,
        protocolContext,
        transactionPool,
        miningConfiguration,
        syncState,
        new BackwardSyncContext(
            protocolContext,
            protocolSchedule,
            syncConfig,
            metricsSystem,
            ethProtocolManager.ethContext(),
            syncState,
            BackwardChain.from(
                storageProvider, ScheduleBasedBlockHeaderFunctions.create(protocolSchedule))),
        ethProtocolManager.ethContext().getScheduler());
  }

  @Override
  protected EthProtocolManager createEthProtocolManager(
      final ProtocolContext protocolContext,
      final SynchronizerConfiguration synchronizerConfiguration,
      final TransactionPool transactionPool,
      final EthProtocolConfiguration ethereumWireProtocolConfiguration,
      final EthPeers ethPeers,
      final EthContext ethContext,
      final EthMessages ethMessages,
      final EthScheduler scheduler,
      final List<PeerValidator> peerValidators,
      final Optional<MergePeerFilter> mergePeerFilter,
      final ForkIdManager forkIdManager) {

    var mergeContext = protocolContext.getConsensusContext(MergeContext.class);

    var mergeBestPeerComparator =
        new TransitionBestPeerComparator(
            genesisConfigOptions.getTerminalTotalDifficulty().map(Difficulty::of).orElseThrow());
    ethPeers.setBestPeerComparator(mergeBestPeerComparator);
    mergeContext.observeNewIsPostMergeState(mergeBestPeerComparator);

    Optional<MergePeerFilter> filterToUse = Optional.of(new MergePeerFilter());

    if (mergePeerFilter.isPresent()) {
      filterToUse = mergePeerFilter;
    }
    mergeContext.observeNewIsPostMergeState(filterToUse.get());
    mergeContext.addNewUnverifiedForkchoiceListener(filterToUse.get());

    EthProtocolManager ethProtocolManager =
        super.createEthProtocolManager(
            protocolContext,
            synchronizerConfiguration,
            transactionPool,
            ethereumWireProtocolConfiguration,
            ethPeers,
            ethContext,
            ethMessages,
            scheduler,
            peerValidators,
            filterToUse,
            forkIdManager);

    return ethProtocolManager;
  }

  /**
   * Create transition mining coordinator.
   *
   * @param protocolSchedule the protocol schedule
   * @param protocolContext the protocol context
   * @param transactionPool the transaction pool
   * @param miningConfiguration the mining parameters
   * @param syncState the sync state
   * @param backwardSyncContext the backward sync context
   * @param ethScheduler the scheduler
   * @return the mining coordinator
   */
  protected MiningCoordinator createTransitionMiningCoordinator(
      final ProtocolSchedule protocolSchedule,
      final ProtocolContext protocolContext,
      final TransactionPool transactionPool,
      final MiningConfiguration miningConfiguration,
      final SyncState syncState,
      final BackwardSyncContext backwardSyncContext,
      final EthScheduler ethScheduler) {

    this.syncState.set(syncState);

    final Optional<Address> depositContractAddress =
        genesisConfigOptions.getDepositContractAddress();

    return new MergeCoordinator(
        protocolContext,
        protocolSchedule,
        ethScheduler,
        transactionPool,
        miningConfiguration,
        backwardSyncContext,
        depositContractAddress);
  }

  @Override
  protected ProtocolSchedule createProtocolSchedule() {
    return MergeProtocolSchedule.create(
        genesisConfigOptions,
        privacyParameters,
        isRevertReasonEnabled,
        miningConfiguration,
        badBlockManager,
        isParallelTxProcessingEnabled,
        metricsSystem);
  }

  @Override
  protected MergeContext createConsensusContext(
      final Blockchain blockchain,
      final WorldStateArchive worldStateArchive,
      final ProtocolSchedule protocolSchedule) {

    final OptionalLong terminalBlockNumber = genesisConfigOptions.getTerminalBlockNumber();
    final Optional<Hash> terminalBlockHash = genesisConfigOptions.getTerminalBlockHash();
    final boolean isPostMergeAtGenesis =
        genesisConfigOptions.getTerminalTotalDifficulty().isPresent()
            && genesisConfigOptions.getTerminalTotalDifficulty().get().isZero()
            && blockchain.getGenesisBlockHeader().getDifficulty().isZero();

    final MergeContext mergeContext =
        PostMergeContext.get()
            .setSyncState(syncState.get())
            .setTerminalTotalDifficulty(
                genesisConfigOptions
                    .getTerminalTotalDifficulty()
                    .map(Difficulty::of)
                    .orElse(Difficulty.ZERO))
            .setPostMergeAtGenesis(isPostMergeAtGenesis);

    blockchain
        .getFinalized()
        .flatMap(blockchain::getBlockHeader)
        .ifPresent(mergeContext::setFinalized);

    blockchain
        .getSafeBlock()
        .flatMap(blockchain::getBlockHeader)
        .ifPresent(mergeContext::setSafeBlock);

    if (terminalBlockNumber.isPresent() && terminalBlockHash.isPresent()) {
      Optional<BlockHeader> termBlock = blockchain.getBlockHeader(terminalBlockNumber.getAsLong());
      mergeContext.setTerminalPoWBlock(termBlock);
    }
    blockchain.observeBlockAdded(
        blockAddedEvent ->
            blockchain
                .getTotalDifficultyByHash(blockAddedEvent.getBlock().getHeader().getHash())
                .ifPresent(mergeContext::setIsPostMerge));

    return mergeContext;
  }

  @Override
  protected PluginServiceFactory createAdditionalPluginServices(
      final Blockchain blockchain, final ProtocolContext protocolContext) {
    return new NoopPluginServiceFactory();
  }

  @Override
  protected List<PeerValidator> createPeerValidators(
      final ProtocolSchedule protocolSchedule, final PeerTaskExecutor peerTaskExecutor) {
    List<PeerValidator> retval = super.createPeerValidators(protocolSchedule, peerTaskExecutor);
    final OptionalLong powTerminalBlockNumber = genesisConfigOptions.getTerminalBlockNumber();
    final Optional<Hash> powTerminalBlockHash = genesisConfigOptions.getTerminalBlockHash();
    if (powTerminalBlockHash.isPresent() && powTerminalBlockNumber.isPresent()) {
      retval.add(
          new RequiredBlocksPeerValidator(
              protocolSchedule,
              peerTaskExecutor,
              syncConfig,
              metricsSystem,
              powTerminalBlockNumber.getAsLong(),
              powTerminalBlockHash.get(),
              0));
    } else {
      LOG.debug("unable to validate peers with terminal difficulty blocks");
    }
    return retval;
  }

  @Override
  public BesuController build() {
    final BesuController controller = super.build();
    PostMergeContext.get().setSyncState(controller.getSyncState());
    return controller;
  }
}

File: besu/src/main/java/org/hyperledger/besu/controller/MergeBesuControllerBuilder.kt
/*
 * Copyright contributors to Hyperledger Besu.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package org.hyperledger.besu.controller

import org.apache.tuweni.units.bigints.UInt256
import org.hyperledger.besu.consensus.merge.MergeContext
import org.hyperledger.besu.consensus.merge.MergeProtocolSchedule
import org.hyperledger.besu.consensus.merge.PostMergeContext
import org.hyperledger.besu.consensus.merge.TransitionBestPeerComparator
import org.hyperledger.besu.consensus.merge.blockcreation.MergeCoordinator
import org.hyperledger.besu.datatypes.Hash
import org.hyperledger.besu.ethereum.ProtocolContext
import org.hyperledger.besu.ethereum.blockcreation.MiningCoordinator
import org.hyperledger.besu.ethereum.chain.BlockAddedEvent
import org.hyperledger.besu.ethereum.chain.Blockchain
import org.hyperledger.besu.ethereum.core.BlockHeader
import org.hyperledger.besu.ethereum.core.Difficulty
import org.hyperledger.besu.ethereum.core.MiningConfiguration
import org.hyperledger.besu.ethereum.eth.EthProtocolConfiguration
import org.hyperledger.besu.ethereum.eth.manager.*
import org.hyperledger.besu.ethereum.eth.manager.peertask.PeerTaskExecutor
import org.hyperledger.besu.ethereum.eth.peervalidation.PeerValidator
import org.hyperledger.besu.ethereum.eth.peervalidation.RequiredBlocksPeerValidator
import org.hyperledger.besu.ethereum.eth.sync.SynchronizerConfiguration
import org.hyperledger.besu.ethereum.eth.sync.backwardsync.BackwardChain
import org.hyperledger.besu.ethereum.eth.sync.backwardsync.BackwardSyncContext
import org.hyperledger.besu.ethereum.eth.sync.state.SyncState
import org.hyperledger.besu.ethereum.eth.transactions.TransactionPool
import org.hyperledger.besu.ethereum.forkid.ForkIdManager
import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule
import org.hyperledger.besu.ethereum.mainnet.ScheduleBasedBlockHeaderFunctions
import org.hyperledger.besu.ethereum.worldstate.WorldStateArchive
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.atomic.AtomicReference

/** The Merge besu controller builder.  */
class MergeBesuControllerBuilder
/** Default constructor.  */
    : BesuControllerBuilder() {
    private val syncState = AtomicReference<SyncState?>()

    /**
     * Gets post merge context.
     *
     * @return the post merge context
     */
    @JvmField
    val postMergeContext: PostMergeContext = PostMergeContext()

    override fun createMiningCoordinator(
        protocolSchedule: ProtocolSchedule?,
        protocolContext: ProtocolContext?,
        transactionPool: TransactionPool?,
        miningConfiguration: MiningConfiguration?,
        syncState: SyncState?,
        ethProtocolManager: EthProtocolManager?
    ): MiningCoordinator {
        return createTransitionMiningCoordinator(
            protocolSchedule,
            protocolContext!!,
            transactionPool,
            miningConfiguration!!,
            syncState,
            BackwardSyncContext(
                protocolContext,
                protocolSchedule,
                syncConfig,
                metricsSystem,
                ethProtocolManager!!.ethContext(),
                syncState,
                BackwardChain.from(
                    storageProvider, ScheduleBasedBlockHeaderFunctions.create(protocolSchedule)
                )
            ),
            ethProtocolManager.ethContext().scheduler
        )
    }

    override fun createEthProtocolManager(
        protocolContext: ProtocolContext,
        synchronizerConfiguration: SynchronizerConfiguration?,
        transactionPool: TransactionPool?,
        ethereumWireProtocolConfiguration: EthProtocolConfiguration,
        ethPeers: EthPeers?,
        ethContext: EthContext?,
        ethMessages: EthMessages?,
        scheduler: EthScheduler?,
        peerValidators: List<PeerValidator>?,
        mergePeerFilter: Optional<MergePeerFilter>,
        forkIdManager: ForkIdManager?
    ): EthProtocolManager {
        val mergeContext = protocolContext!!.getConsensusContext(MergeContext::class.java)

        val mergeBestPeerComparator =
            TransitionBestPeerComparator(
                genesisConfigOptions!!.terminalTotalDifficulty.map { value: UInt256? -> Difficulty.of(value) }
                    .orElseThrow()
            )
        ethPeers!!.bestPeerComparator = mergeBestPeerComparator
        mergeContext.observeNewIsPostMergeState(mergeBestPeerComparator)

        var filterToUse = Optional.of(MergePeerFilter())

        if (mergePeerFilter!!.isPresent) {
            filterToUse = mergePeerFilter
        }
        mergeContext.observeNewIsPostMergeState(filterToUse.get())
        mergeContext.addNewUnverifiedForkchoiceListener(filterToUse.get())

        val ethProtocolManager =
            super.createEthProtocolManager(
                protocolContext!!,
                synchronizerConfiguration,
                transactionPool,
                ethereumWireProtocolConfiguration,
                ethPeers,
                ethContext,
                ethMessages,
                scheduler,
                peerValidators,
                filterToUse,
                forkIdManager
            )

        return ethProtocolManager
    }

    /**
     * Create transition mining coordinator.
     *
     * @param protocolSchedule the protocol schedule
     * @param protocolContext the protocol context
     * @param transactionPool the transaction pool
     * @param miningConfiguration the mining parameters
     * @param syncState the sync state
     * @param backwardSyncContext the backward sync context
     * @param ethScheduler the scheduler
     * @return the mining coordinator
     */
    fun createTransitionMiningCoordinator(
        protocolSchedule: ProtocolSchedule?,
        protocolContext: ProtocolContext,
        transactionPool: TransactionPool?,
        miningConfiguration: MiningConfiguration,
        syncState: SyncState?,
        backwardSyncContext: BackwardSyncContext?,
        ethScheduler: EthScheduler?
    ): MiningCoordinator {
        this.syncState.set(syncState)

        val depositContractAddress =
            genesisConfigOptions!!.depositContractAddress

        return MergeCoordinator(
            protocolContext,
            protocolSchedule,
            ethScheduler,
            transactionPool,
            miningConfiguration,
            backwardSyncContext,
            depositContractAddress
        )
    }

    override fun createProtocolSchedule(): ProtocolSchedule {
        return MergeProtocolSchedule.create(
            genesisConfigOptions,
            privacyParameters,
            isRevertReasonEnabled,
            miningConfiguration,
            badBlockManager,
            isParallelTxProcessingEnabled,
            metricsSystem
        )
    }

    override fun createConsensusContext(
        blockchain: Blockchain?,
        worldStateArchive: WorldStateArchive?,
        protocolSchedule: ProtocolSchedule?
    ): MergeContext? {
        val terminalBlockNumber = genesisConfigOptions!!.terminalBlockNumber
        val terminalBlockHash = genesisConfigOptions!!.terminalBlockHash
        val isPostMergeAtGenesis =
            genesisConfigOptions!!.terminalTotalDifficulty.isPresent
                    && genesisConfigOptions!!.terminalTotalDifficulty.get().isZero
                    && blockchain!!.genesisBlockHeader.difficulty.isZero

        val mergeContext: MergeContext =
            postMergeContext
                .setSyncState(syncState.get())
                .setTerminalTotalDifficulty(
                    genesisConfigOptions!!
                        .getTerminalTotalDifficulty()
                        .map { value: UInt256? -> Difficulty.of(value) }
                        .orElse(Difficulty.ZERO))
                .setPostMergeAtGenesis(isPostMergeAtGenesis)

        blockchain!!
            .finalized
            .flatMap { blockHeaderHash: Hash? ->
                blockchain.getBlockHeader(
                    blockHeaderHash
                )
            }
            .ifPresent { blockHeader: BlockHeader? ->
                mergeContext.setFinalized(
                    blockHeader
                )
            }

        blockchain
            .safeBlock
            .flatMap { blockHeaderHash: Hash? ->
                blockchain.getBlockHeader(
                    blockHeaderHash
                )
            }
            .ifPresent { blockHeader: BlockHeader? ->
                mergeContext.setSafeBlock(
                    blockHeader
                )
            }

        if (terminalBlockNumber.isPresent && terminalBlockHash.isPresent) {
            val termBlock = blockchain.getBlockHeader(terminalBlockNumber.asLong)
            mergeContext.terminalPoWBlock = termBlock
        }
        blockchain.observeBlockAdded { blockAddedEvent: BlockAddedEvent ->
            blockchain
                .getTotalDifficultyByHash(blockAddedEvent.block.header.hash)
                .ifPresent { totalDifficulty: Difficulty? -> mergeContext.setIsPostMerge(totalDifficulty) }
        }

        return mergeContext
    }

    override fun createAdditionalPluginServices(
        blockchain: Blockchain?, protocolContext: ProtocolContext?
    ): PluginServiceFactory {
        return NoopPluginServiceFactory()
    }

    override fun createPeerValidators(
        protocolSchedule: ProtocolSchedule?, peerTaskExecutor: PeerTaskExecutor?
    ): List<PeerValidator> {
        val retval = super.createPeerValidators(protocolSchedule, peerTaskExecutor)
        val powTerminalBlockNumber = genesisConfigOptions!!.terminalBlockNumber
        val powTerminalBlockHash = genesisConfigOptions!!.terminalBlockHash
        if (powTerminalBlockHash.isPresent && powTerminalBlockNumber.isPresent) {
            retval.toMutableList().add(
                RequiredBlocksPeerValidator(
                    protocolSchedule,
                    peerTaskExecutor,
                    syncConfig,
                    metricsSystem,
                    powTerminalBlockNumber.asLong,
                    powTerminalBlockHash.get(),
                    0
                )
            )
        } else {
            LOG.debug("unable to validate peers with terminal difficulty blocks")
        }
        return retval
    }

    override fun build(): BesuController? {
        val controller = super.build()
        postMergeContext.setSyncState(syncState.get())
        return controller
    }

    companion object {
        private val LOG: Logger = LoggerFactory.getLogger(MergeBesuControllerBuilder::class.java)
    }
}
